src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -0,0 +1,397 @@
+    {
+        if (fabric == fabricIndex)
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    mFabricList.push_back(fabricIndex);
+
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    size_t counter      = mFabricList.size() + 1;
+    size_t total = FABRIC_INDEX_TLV_SIZE * counter;
+    backingBuffer.Calloc(total);
+    ReturnErrorCodeIf(backingBuffer.Get() == nullptr, CHIP_ERROR_NO_MEMORY);
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), total);
+
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, arrayType));
+    for (auto & fabric : mFabricList)
+    {
+        ReturnErrorOnFailure(writer.Put(TLV::AnonymousTag(), fabric));
+    }
+    return writer.EndContainer(arrayType);
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+    ReturnErrorOnFailure(mpClientInfoStore->SyncSetKeyValue(DefaultStorageKeyAllocator::ICDFabricList().KeyName(),
+                                                                     backingBuffer.Get(), static_cast<uint16_t>(len)));
+    uint16_t counterLen = static_cast<uint16_t>(sizeof(counter));
+    ReturnErrorOnFailure(mpClientInfoStore->SyncDeleteKeyValue(
+        DefaultStorageKeyAllocator::ICDFabricListCounter().KeyName()));
+    return mpClientInfoStore->SyncSetKeyValue(
+        DefaultStorageKeyAllocator::ICDFabricListCounter().KeyName(), &counter, counterLen);
+}
+
+CHIP_ERROR DefaultICDClientStorage::LoadFabricList()
+{
+    size_t counter      = 0;
+    uint16_t counterLen = static_cast<uint16_t>(sizeof(counter));
+    ReturnErrorOnFailure(mpClientInfoStore->SyncGetKeyValue(
+        DefaultStorageKeyAllocator::ICDFabricListCounter().KeyName(), &counter, counterLen));
+
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    size_t len = FABRIC_INDEX_TLV_SIZE * counter;
+    ReturnErrorCodeIf(!backingBuffer.Calloc(len), CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+    uint16_t length = static_cast<uint16_t>(len);
+    ReturnErrorOnFailure(mpClientInfoStore->SyncGetKeyValue(DefaultStorageKeyAllocator::ICDFabricList().KeyName(),
+                                                                    backingBuffer.Get(), length));
+
+    TLV::ScopedBufferTLVReader reader(std::move(backingBuffer), len);
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(reader.EnterContainer(arrayType));
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    while ((err = reader.Next(TLV::kTLVType_UnsignedInteger, TLV::AnonymousTag())) == CHIP_NO_ERROR)
+    {
+        FabricIndex fabricIndex;
+        ReturnErrorOnFailure(reader.Get(fabricIndex));
+        mFabricList.push_back(fabricIndex);
+    }
+
+    if (err != CHIP_END_OF_TLV)
```

Comment:
It should not be an error for there to be more members than expected. This can happen after a logical downgrade after a firmware update. There only needs to be the reading of elements understood, and the rest is left alone.

Reply:
thanks, updated